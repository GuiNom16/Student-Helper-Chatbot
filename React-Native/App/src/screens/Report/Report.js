import {
  View,
  Text,
  TouchableOpacity,
  StyleSheet,
  TextInput,
  Alert,
  Image,
  Modal,
} from "react-native";

import { useState } from "react";
import { Dropdown } from "react-native-element-dropdown";
import { db } from "../../firebase";
import { addDoc, collection } from "firebase/firestore";

import LottieView from "lottie-react-native";

const data = [
  { label: "JoeBot", value: "1" },
  { label: "QueryBot", value: "2" },
  { label: "Both", value: "3" },
];

const Report = () => {
  const [isSelected, setSelection] = useState("");
  const [description, setDescription] = useState("");
  const [modalVisible, setModalVisible] = useState(false);
  const [value, setValue] = useState(null);
  const [isFocus, setIsFocus] = useState(false);

  //add a new field
  const addReport = () => {
    //check if we have new field data
    if (isSelected != "") {
      addDoc(collection(db, "reports"), {
        classification: isSelected,
        description: description,
        botType: value,
      }).then(
        () => setDescription(""),
        setSelection(""),
        setValue(null),
        setModalVisible(!modalVisible)
      );
    }
  };

  return (
    <View style={styles.container}>
      <Modal
        animationType="slide"
        transparent={true}
        visible={modalVisible}
        onRequestClose={() => {
          Alert.alert("Popup has been closed.");
          setModalVisible(!modalVisible);
        }}
      >
        <View style={styles.centeredView}>
          <View style={styles.modalView}>
            <LottieView
              source={require("../../lottie/feedback.json")}
              autoPlay
              style={{ height: 100, width: 100 }}
            />
            <Text style={styles.modalText}>Report Sent!</Text>
            <TouchableOpacity
              style={[styles.buttonClose]}
              onPress={() => setModalVisible(!modalVisible)}
            >
              <Text style={styles.textStyle}>Close</Text>
            </TouchableOpacity>
          </View>
        </View>
      </Modal>

      <Text style={{ fontSize: 18, fontWeight: "bold" }}>
        Help us get better at what we do
      </Text>
      <View style={styles.reportContainer}>
        <Dropdown
          style={[styles.dropdown, isFocus && { borderColor: "blue" }]}
          placeholderStyle={styles.placeholderStyle}
          selectedTextStyle={styles.selectedTextStyle}
          iconStyle={styles.iconStyle}
          data={data}
          maxHeight={300}
          labelField="label"
          valueField="value"
          placeholder={!isFocus ? "Pick problematic chatbot" : "..."}
          value={value}
          onFocus={() => setIsFocus(true)}
          onBlur={() => setIsFocus(false)}
          onChange={(item) => {
            setValue(item.value);
            setIsFocus(false);
          }}
          renderLeftIcon={() => (
            <Image
              source={require("../../../assets/images/bot.png")}
              style={{ height: 35, width: 35, marginRight: 5 }}
              color={isFocus ? "blue" : "black"}
              name="Safety"
              size={20}
            />
          )}
        />

        <Text style={{ marginTop: 10 }}>Type:</Text>
        <TouchableOpacity onPress={() => setSelection("Off Topic")}>
          <Text
            style={isSelected == "Off Topic" ? styles.active : styles.unactive}
          >
            Off Topic Response
          </Text>
        </TouchableOpacity>
        <TouchableOpacity onPress={() => setSelection("Misleading")}>
          <Text
            style={isSelected == "Misleading" ? styles.active : styles.unactive}
          >
            Misleading or Deceiving response
          </Text>
        </TouchableOpacity>
        <TouchableOpacity onPress={() => setSelection("Other")}>
          <Text style={isSelected == "Other" ? styles.active : styles.unactive}>
            Other
          </Text>
        </TouchableOpacity>
        <Text style={{ marginTop: 10 }}>Description:</Text>
        <TextInput
          style={styles.TextInput}
          onChangeText={setDescription}
          value={description}
          placeholder="Explain issue in more details"
          multiline={true}
        ></TextInput>
      </View>

      <TouchableOpacity
        onPress={() => {
          if (isSelected != "" && value != null && description != "") {
            Alert.alert(
              "Report Submission",
              "Are you sure you want to submit your feedback?",
              [
                {
                  text: "Not now",
                  onPress: () => console.log("Cancel Pressed"),
                  style: "cancel",
                },
                { text: "Yes", onPress: () => addReport() },
              ]
            );
          } else {
            Alert.alert(
              "Opps!",
              "Make sure that all the fields are filled before submitting a report",
              [
                {
                  text: "Close",
                  onPress: () => console.log("Cancel Pressed"),
                  style: "cancel",
                },
              ]
            );
          }
        }}
        style={styles.buttonFeedback}
      >
        <Text style={styles.buttonFeedbackText}>Send Feedback</Text>
      </TouchableOpacity>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: "#d3d3d3",
    alignItems: "center",
    justifyContent: "center",
  },
  buttonFeedback: {
    color: "white",
    width: "65%",
    borderRadius: 25,
    height: 50,

    alignItems: "center",
    justifyContent: "center",
    backgroundColor: "#0474ea",
  },
  buttonFeedbackText: {
    color: "white",
    fontWeight: "bold",
    fontSize: 18,
  },
  reportContainer: {
    backgroundColor: "white",
    height: 500,
    width: 320,
    margin: 30,
    borderRadius: 20,
    padding: 20,
  },
  TextInput: {
    backgroundColor: "#d3d3d3",
    textAlignVertical: "top",
    height: 180,
    padding: 10,
    marginTop: 10,
    borderRadius: 20,
  },
  active: {
    backgroundColor: "black",
    color: "#d3d3d3",
    padding: 10,
    borderRadius: 10,
    marginTop: 10,
  },
  unactive: {
    backgroundColor: "#d3d3d3",
    color: "black",
    padding: 10,
    borderRadius: 10,
    marginTop: 10,
  },
  dropdown: {
    height: 50,
    backgroundColor: "#d3d3d3",
    borderColor: "gray",
    borderWidth: 0.5,
    borderRadius: 8,
    paddingHorizontal: 8,
  },
  centeredView: {
    flex: 1,
    justifyContent: "center",
    alignItems: "center",
    marginTop: 22,
  },
  modalView: {
    margin: 20,
    backgroundColor: "white",
    borderRadius: 20,
    padding: 35,
    alignItems: "center",
    shadowColor: "#000",
    shadowOffset: {
      width: 0,
      height: 2,
    },
    shadowOpacity: 0.25,
    shadowRadius: 4,
    elevation: 5,
  },
  button: {
    height: 50,
    width: 120,
    borderRadius: 10,
    padding: 10,
    elevation: 2,
    marginRight: 25,
  },
  modalText: {
    marginBottom: 15,
    textAlign: "center",
    color: "white",
  },
  buttonClose: {
    borderRadius: 10,
    padding: 12,
    backgroundColor: "#2196F3",
  },
});

export default Report;
