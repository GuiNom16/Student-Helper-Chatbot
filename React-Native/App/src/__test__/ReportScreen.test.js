import React from "react";
import renderer from "react-test-renderer";

import Report from "../screens/Report/Report";

jest.mock("@react-navigation/native", () => {
  const actualNav = jest.requireActual("@react-navigation/native");
  return {
    ...actualNav,
    useNavigation: () => ({
      navigate: jest.fn(),
      dispatch: jest.fn(),
    }),
  };
});

//testing if the right amount of children is being displayed
describe("Report screen", () => {
  const tree = renderer.create(<Report />).toJSON();
  it("has 4 children", () => {
    expect(tree.children.length).toBe(4);
  });

  it("screen renders correctly", () => {
    expect(tree).toMatchSnapshot();
  });
});
