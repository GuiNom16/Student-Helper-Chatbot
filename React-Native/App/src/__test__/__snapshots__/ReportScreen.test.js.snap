// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Report screen screen renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#d3d3d3",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Modal
    animationType="slide"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    transparent={true}
    visible={false}
  />
  <Text
    style={
      Object {
        "fontSize": 18,
        "fontWeight": "bold",
      }
    }
  >
    Help us get better at what we do
  </Text>
  <View
    style={
      Object {
        "backgroundColor": "white",
        "borderRadius": 20,
        "height": 500,
        "margin": 30,
        "padding": 20,
        "width": 320,
      }
    }
  >
    <View
      onLayout={[Function]}
      style={
        Object {
          "backgroundColor": "#d3d3d3",
          "borderColor": "gray",
          "borderRadius": 8,
          "borderWidth": 0.5,
          "height": 50,
          "justifyContent": "center",
          "paddingHorizontal": 8,
        }
      }
    >
      <View
        accessible={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "height": 35,
            "justifyContent": "space-between",
          }
        }
      >
        <Image
          color="black"
          name="Safety"
          size={20}
          source={1}
          style={
            Object {
              "height": 35,
              "marginRight": 5,
              "width": 35,
            }
          }
        />
        <Text
          style={
            Array [
              Object {
                "flex": 1,
                "fontSize": 16,
                "writingDirection": "ltr",
              },
              undefined,
              Object {},
            ]
          }
        >
          Pick problematic chatbot
        </Text>
        <Image
          source={1}
          style={
            Object {
              "height": 20,
              "tintColor": "gray",
              "width": 20,
            }
          }
        />
      </View>
    </View>
    <Text
      style={
        Object {
          "marginTop": 10,
        }
      }
    >
      Type:
    </Text>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <Text
        style={
          Object {
            "backgroundColor": "#d3d3d3",
            "borderRadius": 10,
            "color": "black",
            "marginTop": 10,
            "padding": 10,
          }
        }
      >
        Off Topic Response
      </Text>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <Text
        style={
          Object {
            "backgroundColor": "#d3d3d3",
            "borderRadius": 10,
            "color": "black",
            "marginTop": 10,
            "padding": 10,
          }
        }
      >
        Misleading or Deceiving response
      </Text>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <Text
        style={
          Object {
            "backgroundColor": "#d3d3d3",
            "borderRadius": 10,
            "color": "black",
            "marginTop": 10,
            "padding": 10,
          }
        }
      >
        Other
      </Text>
    </View>
    <Text
      style={
        Object {
          "marginTop": 10,
        }
      }
    >
      Description:
    </Text>
    <TextInput
      multiline={true}
      onChangeText={[Function]}
      placeholder="Explain issue in more details"
      style={
        Object {
          "backgroundColor": "#d3d3d3",
          "borderRadius": 20,
          "height": 180,
          "marginTop": 10,
          "padding": 10,
          "textAlignVertical": "top",
        }
      }
      value=""
    />
  </View>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#0474ea",
        "borderRadius": 25,
        "color": "white",
        "height": 50,
        "justifyContent": "center",
        "opacity": 1,
        "width": "65%",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "white",
          "fontSize": 18,
          "fontWeight": "bold",
        }
      }
    >
      Send Feedback
    </Text>
  </View>
</View>
`;
